File: Merge_Two_Binary_Trees/Answer.java
package mtbt;

import java.util.ArrayList;
import java.util.List;

/**
 * Definition for a binary tree node.
 */

class TreeNode {

    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

class Solution {
    public TreeNode mergeTrees(TreeNode t1, TreeNode t2) {

        if(t1 == null) {
            return t2;
        }
        if(t2 == null) {
            return t1;
        }

        t1.val = t1.val + t2.val;
        t1.left = mergeTrees(t1.left, t2.left);
        t1.right = mergeTrees(t1.right, t2.right);
        return t1;
    }
}

public class Answer {
    public static void main(String[] args) {

        int[] arr1 = { 1, 3, 2, 5 };
        int[] arr2 = { 2, 1, 3, 8, 4, 6, 7 };

        TreeNode t1 = createTree(arr1);
        TreeNode t2 = createTree(arr2);

        Solution sol = new Solution();
        TreeNode ret = sol.mergeTrees(t1, t2);

        System.out.println(ret);
    }

    static TreeNode createTree(int[] arr) {

        List<TreeNode> tree = new ArrayList<>();
        for (int i = 0; i < arr.length; i++) {
            TreeNode node = new TreeNode(arr[i]);
            tree.add(node);
        }

        // 获取树的层级(不包括叶子节点)
        int hierarchy = arr.length / 2 - 1;

        // 从根开始设置左右叶子节点
        for (int parentIndex = 0; parentIndex < hierarchy; parentIndex++) {
            tree.get(parentIndex).left = tree.get(parentIndex * 2 + 1);
            tree.get(parentIndex).right = tree.get(parentIndex * 2 + 2);
        }

        // 最后一个父节点
        tree.get(hierarchy).left = tree.get(hierarchy * 2 + 1);
        if(arr.length % 2 == 1) {
            tree.get(hierarchy).right = tree.get(hierarchy * 2 + 2);
        }
        return tree.get(0);
    }
}

File: Merge_Two_Binary_Trees/answer.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

import math


class TreeNode:

    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    @staticmethod
    def createTree(ary) -> TreeNode:

        treenode = []
        for elem in ary:
            treenode.append(TreeNode(elem))

        aryLen = len(ary)
        hierarchy = aryLen / 2 - 1

        for index in range(math.ceil(hierarchy)):
            treenode[index].left = treenode[index * 2 + 1]
            treenode[index].right = treenode[index * 2 + 2]

        hierarchy = math.floor(hierarchy)
        treenode[hierarchy].left = treenode[hierarchy * 2 + 1]
        if aryLen % 2 == 1:
            treenode[hierarchy].right = treenode[hierarchy * 2 + 2]

        return treenode[0]

    def mergeTrees(self, t1: TreeNode, t2: TreeNode) -> TreeNode:

        if t1 is not None and t2 is not None:

            node = TreeNode(t1.val + t2.val)
            node.left = self.mergeTrees(t1.left, t2.left)
            node.right = self.mergeTrees(t1.right, t2.right)

            return node
        elif t1 is None:
            return t2
        elif t2 is None:
            return t1


ary1 = [1, 3, 2, 5]
ary2 = [2, 1, 3, 8, 4, 6, 7]

sol = Solution()
t1 = sol.createTree(ary1)
t2 = sol.createTree(ary2)
print(sol.mergeTrees(t1, t2))
